### PR TITLE
Introduce a formal policy for maintaining cloudproviders

### PR DESCRIPTION
**Pull request by MaciekPytel**
This is an initial proposal for a policy on cloudprovider maintenance discussed in a recent sig-meeting.

The policy largely codifies what we've already been doing for years (including the requirements we've already imposed on new providers). It also introduces a new 'cloudprovider maintenance request' mechanism, following the general idea previously discussed in the sig-meeting.

#### Which component this PR applies to?

cluster-autoscaler

#### What type of PR is this?

/kind documentation

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

This policy is only relevant for CA / cloudprovider developers and not its users.

### CODE REVIEW

**Comment by MaciekPytel** [comment]
/hold 
Holding to avoid any accidental merge before the discussion is done.

sig-leads for approval: @gjtempleton @mwielgus 

cc: @ericrrath @jackfrancis @piepmatz @gajicdev @7fELF @ringtail @jtherin @elmiko @Andrius521 @nilo19 @ellistarn @gjtempleton @davidjumani @birdiesanders @DMajrekar @kevin-wangzefeng @dbonfigli @ddymko @thirdeyenick @displague @andrewsykim @zalmarge @schegi @enxebre @d-mo @alphajc @pawelkuc @RealHarshThakur @towca @jaypipes @drmorr0 @avorima @detiber @Sh4d1 @feiskyer @vishalanarase @PhilippeChepy @Jeffwan @OriHoch @louisportay @hardikdr @jlamillan @hello2mao @pierre-emmanuelJ @x13n @mrajashree @tghartland @remyleone @shysank @RainbowMango @LKaemmerling @ctrox @cprivitere @4ND3R50N @v-pap @ArturasRa @deitch @arunmk @aleksandra-malinowska @MorrisLaw @marwanad

**Comment by MaciekPytel** [comment]
At least the github UI doesn't seem to be rendering some of the logins in the comment above correctly (is there maximum logins per line limit?) and I'm not sure if the notifications went out. Just in case:

cc: @RainbowMango @LKaemmerling @ctrox @cprivitere @4ND3R50N @v-pap @ArturasRa @deitch @arunmk @aleksandra-malinowska @MorrisLaw @marwanad

**Comment by deitch** [comment]
I don't think I got the alert for this, so thanks for @ cc'ing after.

This all looks good, glad to see it somewhat formalized.

The out-of-tree provider appears to be almost an after-thought, or a fallback. Is there any desire to move it entirely to out-of-tree, the same way, e.g. CCM is done?

**Comment by MaciekPytel** [comment]
Thanks for confirming my suspicions on notification not working @deitch, I'll make sure to split it into multiple lines next time.

Re: external provider:
 * No plans to remove in-tree providers. There are things you can do in-tree that you can't do over grpc such as implementing custom processors to be used alongside your cloudprovider. Also, as far as I know, no one tried to run externalgrpc provider in a cluster with thousands of nodes and I doubt it would work. For comparison, it took us years to get GCE provider to reliably work at 5k node scale. This is purely a guess, but I'd imagine the proto would need to be updated to allow batching requests and the CA side of implementation would likely need much more caching and optimization. Given that all the very large clusters I know of run on existing in-tree providers, I don't know if anyone will be willing to invest that much effort into optimizing externalgrpc. 
 * Based on the above, my recommendation to anyone building a new provider would be to start with externalgrpc if they just want to get it working with minimal effort. However, if you want some advanced, customized functionality and/or thousands of nodes scale and your're ready to invest a few months of fulltime effort, you're probably better off with in-tree.
 * I think you're right on externalgrpc sounding like a fallback / secondary option in the doc though, which is not what I intended. I'll update the doc to make external the first option. Thanks for the suggestion!

File: cluster-autoscaler/cloudprovider/POLICY.md
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/POLICY.md:65]: is it worth mentioning that cloudproviders are welcome to carry dependencies in their directory if they need?

File: cluster-autoscaler/cloudprovider/POLICY.md
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/POLICY.md:115]: i think it would be nice to bring this part on a separate review since it proposes a new process.

File: cluster-autoscaler/cloudprovider/POLICY.md
Comment (elmiko) [inline cluster-autoscaler/cloudprovider/POLICY.md:127]: i might bring this on a separate review as well, but i think it's so useful to have that we should include in the primary pr.

**Review by elmiko — COMMENTED** [review]
overall looks good to me, i have a couple comments but no major changes.

File: cluster-autoscaler/cloudprovider/POLICY.md
Comment (jaypipes) [inline cluster-autoscaler/cloudprovider/POLICY.md:104]: :+1: seems fair to me.

**Review by jaypipes — APPROVED** [review]
Thanks @MaciekPytel for this. I know I haven't been able to attend many SIG meetings this year and that the AWS cloud provider has a number of open issues/PRs on it. Will try to do better at filling these gaps.

Proposal looks reasonable and fair to me.

File: cluster-autoscaler/cloudprovider/POLICY.md
Comment (jlamillan) [inline cluster-autoscaler/cloudprovider/POLICY.md:82]: Typo. Should read:
> As such it is also not subject to this policy.

not

> As such it is also not a subject to this policy.

**Comment by x13n** [comment]
Could we make it a part of the policy to have non-empty approvers AND reviewers sections? I just sent out https://github.com/kubernetes/autoscaler/pull/5287 to update one of the OWNERS files, but if we require it, it should be here.

File: cluster-autoscaler/cloudprovider/POLICY.md
Comment (MaciekPytel) [reply cluster-autoscaler/cloudprovider/POLICY.md:89]: done

File: cluster-autoscaler/cloudprovider/POLICY.md
Comment (MaciekPytel) [reply cluster-autoscaler/cloudprovider/POLICY.md]: fixed

**Comment by MaciekPytel** [comment]
@x13n, @deitch - addressed your comments. I've also split the CMR part into a separate PR as per @elmiko comment. This one is now essentially writing down what we've been doing for a while anyway.

**Review by elmiko — COMMENTED** [review]
/lgtm

thanks @MaciekPytel

**Comment by x13n** [comment]
Can we cancel the hold on this one now?

**Pull request merged**
